src/replication/replication_node_manager.cpp

```diff
@@ -61,44 +75,75 @@ namespace cubreplication
 
     void finalize ()
     {
+      wait_ha_tasks ();
+
       g_hostname.clear ();
 
       delete g_slave_node;
       g_slave_node = NULL;
       delete g_master_node;
       g_master_node = NULL;
 
-      // stream and stream file are interdependent, therefore first stop the stream
+      // we need to first stop the stream before destroying stream_file
       g_stream->stop ();
       delete g_stream_file;
       g_stream_file = NULL;
       delete g_stream;
       g_stream = NULL;
     }
 
-    void commute_to_master_state ()
+    void commute_to_master_state (const std::function<void (void)> &ha_transitions)
     {
-      delete g_slave_node;
-      g_slave_node = NULL;
-
-      if (g_master_node == NULL)
-	{
-	  g_master_node = new master_node (g_hostname.c_str (), g_stream, g_stream_file);
-	}
+      inc_tasks ();
+
+      cubthread::entry_task *promote_task = new cubthread::entry_callable_task ([&ha_transitions] (
+		  cubthread::entry &context)
+      {
+	if (g_slave_node != NULL)
+	  {
+	    g_slave_node->finish_fetch_from_stream ();
+	  }
+	delete g_slave_node;
+	g_slave_node = NULL;
+
+	if (g_master_node == NULL)
+	  {
+	    g_master_node = new master_node (g_hostname.c_str (), g_stream, g_stream_file);
+	  }
+
+	ha_transitions ();
+	g_commute_cv.notify_all ();
+      }, true);
+
+      auto wp = cubthread::internal_tasks_worker_pool::get_instance ();
+      cubthread::get_manager ()->push_task (wp, promote_task);
     }
 
-    void commute_to_slave_state ()
+    void commute_to_slave_state (const std::function<void (void)> &ha_transitions)
```

Comment:
same as commute_to_master_state, only force argument should be provided here and all transitions should be incorporated in demote_task.